server/src/main/java/org/elasticsearch/action/search/TransportSearchShardsAction.java

```diff
@@ -0,0 +1,164 @@
+import java.util.function.BiFunction;
+
+public class TransportSearchShardsAction extends HandledTransportAction<SearchShardsRequest, SearchShardsResponse> {
+    private final TransportSearchAction transportSearchAction;
+    private final SearchService searchService;
+    private final RemoteClusterService remoteClusterService;
+    private final ClusterService clusterService;
+    private final SearchTransportService searchTransportService;
+    private final NamedWriteableRegistry namedWriteableRegistry;
+
+    @Inject
+    public TransportSearchShardsAction(
+        TransportService transportService,
+        SearchService searchService,
+        ActionFilters actionFilters,
+        ClusterService clusterService,
+        TransportSearchAction transportSearchAction,
+        SearchTransportService searchTransportService,
+        NamedWriteableRegistry namedWriteableRegistry
+    ) {
+        super(SearchShardsAction.NAME, transportService, actionFilters, SearchShardsRequest::new);
+        this.transportSearchAction = transportSearchAction;
+        this.searchService = searchService;
+        this.remoteClusterService = transportService.getRemoteClusterService();
+        this.clusterService = clusterService;
+        this.searchTransportService = searchTransportService;
+        this.namedWriteableRegistry = namedWriteableRegistry;
+    }
+
+    @Override
+    protected void doExecute(Task task, SearchShardsRequest request, ActionListener<SearchShardsResponse> listener) {
+        final long relativeStartNanos = System.nanoTime();
+        SearchRequest original = request.getSearchRequest();
+        final TransportSearchAction.SearchTimeProvider timeProvider = new TransportSearchAction.SearchTimeProvider(
+            original.getOrCreateAbsoluteStartMillis(),
+            relativeStartNanos,
+            System::nanoTime
+        );
+        ActionListener<SearchRequest> rewriteListener = ActionListener.wrap(rewritten -> {
+            final SearchContextId searchContext;
+            final Map<String, OriginalIndices> remoteClusterIndices;
+            if (rewritten.pointInTimeBuilder() != null) {
+                searchContext = rewritten.pointInTimeBuilder().getSearchContextId(namedWriteableRegistry);
+                remoteClusterIndices = TransportSearchAction.getIndicesFromSearchContexts(searchContext, rewritten.indicesOptions());
+            } else {
+                searchContext = null;
+                remoteClusterIndices = remoteClusterService.groupIndices(rewritten.indicesOptions(), rewritten.indices());
+            }
+            OriginalIndices localIndices = remoteClusterIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
+            if (remoteClusterIndices.isEmpty() == false) {
+                throw new IllegalArgumentException("search_shards doesn't support remote indices; request [" + original + "]");
+            }
+            transportSearchAction.executeLocalSearch(
```

Comment:
I see how easy it is to just use TransportSearchAction like this. There's a couple of odd aspects like this is not technically a search, and executeLocalSearch will do search things like adjusting search type, carry around the search executor, it is subject to the `action.search.shard_count.limit` setting. It even decides if it should run the can match phase, which is counter intuitive as it's the whole point of the new action (or do we want to make that optional?). With this I wonder if it's worth the effort to try to extract the useful bits from transport search action and make the two independent. Sharing code would be nice but looks slightly complicated, I'd have to try to see what it looks like.

Reply:
@javanna Thanks for the feedback. I think we can call `CanMatchPreFilterSearchPhase` directly now.